src/HonzaBotner.Services/StandUpStreakService.cs

```diff
@@ -0,0 +1,119 @@
+﻿using System;
+using System.Threading.Tasks;
+using HonzaBotner.Database;
+using HonzaBotner.Services.Contract;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using StandUpStreak = HonzaBotner.Services.Contract.Dto.StandUpStreak;
+
+namespace HonzaBotner.Services;
+
+public class StandUpStreakService : IStandUpStreakService
+{
+    private readonly HonzaBotnerDbContext _dbContext;
+    private readonly ILogger<WarningService> _logger;
+
+    public StandUpStreakService(HonzaBotnerDbContext dbContext, ILogger<WarningService> logger)
+    {
+        _dbContext = dbContext;
+        _logger = logger;
+    }
+
+    public async Task<StandUpStreak?> GetStreak(ulong userId)
+    {
+        Database.StandUpStreak? standUpStreak = await _dbContext.StandUpStreaks
+            .FirstOrDefaultAsync(streak => streak.UserId == userId);
+
+        return standUpStreak == null ? null : GetDto(standUpStreak);
+    }
+
+    public async Task UpdateStreak(ulong userId)
+    {
+        Database.StandUpStreak? streak = await _dbContext.StandUpStreaks
+            .FirstOrDefaultAsync(streak => streak.UserId == userId);
+        ;
+
+        // Create new streak for a new user
+        if (streak is null)
+        {
+            Database.StandUpStreak newStreak = new()
+            {
+                UserId = userId,
+                Freezes = 0,
+                LastDayOfStreak = DateTime.Today.AddDays(-1),
+                Streak = 1,
+                LongestStreak = 1
+            };
+
+            await _dbContext.StandUpStreaks.AddAsync(newStreak);
+
+            try
+            {
+                await _dbContext.SaveChangesAsync();
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, "Couldn't add streak {@Streak}", streak);
+            }
+
+            return;
+        }
+
+        int days = (DateTime.Today.AddDays(-2) - streak.LastDayOfStreak).Days;
+
+        if (days > streak.Freezes) //Streak broken
+        {
+            streak.Freezes = 0;
+            streak.LastDayOfStreak = DateTime.Today.AddDays(-1);
+            streak.Streak = 1;
+        }
+        else //streak restored
```

Comment:
Just a quick nitpick. As I commented above, we don't have any guarantee that we will have each user only once. Actually current implementation calls `UpdateStreak` for each match inside a message. So instead of this else, there needs to be another condition checking if we haven't already changed the date in user's file to today, so that we don't increment streak each time.

Reply:
Check added